### ISSUE TITLE
[BUG] pos_label param in mlflow.sklearn.autolog not working properly

### ISSUE THREAD

**Issue by nelsoncardenas** [mlflow/mlflow#6797]
### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

1.28.0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Monterrey 12.5.1 (chip Apple M1 Pro)
- **Python version**: 3.8.13
- **yarn version, if running the dev UI**: 

### Folder structure

```
./
📄 conda.yml
📁 bug_report/
├── 📄 MLproject
├── 📄 conda.yml
├── 📄 run.py
└── 📄 sample.csv
```

### command to execute the code

`mlflow run -e bug_report/run.py --experiment-name bug_report  .`

### main `conda.yml` file to execute run command:

```yaml
name: env_main
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.10.6
  - pip=22.2.2
  - hydra-core=1.2.0
  - dask=2022.8.1
  - black=22.6.0
  - flake8=5.0.4
  - jupyter=1.0.0
  - scikit-learn=1.1.2
  - seaborn=0.12.0
  - pip:
      - mlflow==1.28.0
      - dvc==2.23.0
      - dvc-s3==2.19.0

```

### bug_report `bug_report/conda.yml` file:

```yaml
name: env_bug_report
channels:
  - conda-forge
  - defaults
dependencies:
  - python=3.8.13
  - pip=22.2.2
  - pandas=1.2.5
  - dask=2021.10.0
  - typer=0.6.1
  - scikit-learn=1.1.2
  - matplotlib=3.5.3
  - pip:
      - mlflow==1.28.0

```

### MLproject

```MLproject
name: bug_report
conda_env: conda.yml

entry_points:
  main:
    parameters:
      path-train:
        type: str

    command: >-
      python run.py --path-train {path-train}

```

```python
import logging

import mlflow.sklearn
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd

logging.basicConfig(level=logging.INFO, format="%(asctime)-15s %(message)s")
logger = logging.getLogger()

##### BAD EXAMPLE #####
# - contains two distinct values in y_train: `Charged off` and `Fully Paid`
# - autolog has parameter `pos_label="Charged Off"`
# - detects pos_label as `Fully Paid`.
logger.info("Bad example")
logger.info("Training model with error in pos_label")

df_train = pd.read_csv("bug_report/sample.csv")
X_train = df_train.drop(columns=["loan_status"])
y_train = np.array(df_train["loan_status"])
logger.info(f"...unique values in y_train {np.unique(y_train)}")
model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Charged Off")
with mlflow.start_run() as _:
    model.fit(X_train, y_train)

##### GOOD EXAMPLE #####
# - contains two distinct values in y_train: `Negative_label` and `Positive_label`
# - autolog has parameter `pos_label="Positive_label"`
# - detects pos_label as `Positive_label`.
logger.info("Good example")
logger.info("Training model without error in pos_label")
X_train_2 = np.array([[1, 0], [0, 1], [1, 0], [1, 1], [1, 1], [1, 2]])
y_train_2 = np.array(
    [
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Positive_label",
        "Positive_label",
    ]
)
logger.info(f"...unique values in y_train {np.unique(y_train_2)}")

model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Positive_label")
with mlflow.start_run() as _:
    model.fit(X_train_2, y_train_2)

```

### `sample.csv` file

I let the data in Google Drive to repeat the mistake.

https://drive.google.com/file/d/1kH-i1_y2B-2vV0KJhpT52vhDs7IpZA3B/view?usp=sharing

### Describe the problem

I use the line `mlflow.sklearn.autolog(pos_label="Charged Off")` to define the positive label (this is the minority class), but the positive label reported in the mlflow server is `"Fully Paid"`.

Evidence:
![image](https://user-images.githubusercontent.com/18086414/190238008-335221c4-965a-48c0-8b7d-03367e66a048.png)

I created dummy samples to detect the problem, but the error didn't happen again. I let the data in Google Drive to repeat the mistake.

https://drive.google.com/file/d/1kH-i1_y2B-2vV0KJhpT52vhDs7IpZA3B/view?usp=sharing

### Code to reproduce issue

```python
import logging

import mlflow.sklearn
from sklearn.linear_model import LogisticRegression
import numpy as np
import pandas as pd

logging.basicConfig(level=logging.INFO, format="%(asctime)-15s %(message)s")
logger = logging.getLogger()

##### BAD EXAMPLE #####
logger.info("Bad example")
logger.info("Training model with error in pos_label")
# - contains two distinct values in y_train: `Charged off` and `Fully Paid`
# - autolog has parameter `pos_label="Charged Off"`
# - detects pos_label as `Fully Paid`.
df_train = pd.read_csv("bug_report/sample.csv")
X_train = df_train.drop(columns=["loan_status"])
y_train = np.array(df_train["loan_status"])
logger.info(f"...unique values in y_train {np.unique(y_train)}")
model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Charged Off")
with mlflow.start_run() as _:
    model.fit(X_train, y_train)

##### GOOD EXAMPLE #####
logger.info("Good example")
logger.info("Training model without error in pos_label")
X_train_2 = np.array([[1, 0], [0, 1], [1, 0], [1, 1], [1, 1], [1, 2]])
y_train_2 = np.array(
    [
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Negative_label",
        "Positive_label",
        "Positive_label",
    ]
)
logger.info(f"...unique values in y_train {np.unique(y_train_2)}")

model = LogisticRegression()

mlflow.sklearn.autolog(pos_label="Positive_label")
with mlflow.start_run() as _:
    model.fit(X_train_2, y_train_2)

```

### Stack trace

```python
# by executing this command mlflow run -e bug_report/run.py --experiment-name bug_repor
```

```python
2022/09/14 14:07:33 INFO mlflow.utils.conda: Conda environment mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77 already exists.
2022/09/14 14:07:33 INFO mlflow.projects.utils: === Created directory /var/folders/nt/w74p_l0x1h1g63lfz0ybhrdr0000gp/T/tmptrl1l7j8 for downloading remote URIs passed to arguments of type 'path' ===
2022/09/14 14:07:33 INFO mlflow.projects.backend.local: === Running command 'source /Users/nelsoncardenas/opt/miniconda3/bin/../etc/profile.d/conda.sh && conda activate mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77 1>&2 && python bug_report/run.py' in run with ID 'ec5922588def43dea40f502f5482245a' === 
2022-09-14 14:07:37,424 Bad example
2022-09-14 14:07:37,424 Training model with error in pos_label
2022-09-14 14:07:37,457 ...unique values in y_train ['Charged Off' 'Fully Paid']
2022/09/14 14:07:44 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/nelsoncardenas/opt/miniconda3/envs/mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77/lib/python3.10/site-packages/mlflow/models/signature.py:129: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details."
2022/09/14 14:07:45 WARNING mlflow.utils.autologging_utils: MLflow autologging encountered a warning: "/Users/nelsoncardenas/opt/miniconda3/envs/mlflow-63e1d0b383727eeea5bafbfa5e13d4e3c832be77/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils."
2022-09-14 14:07:45,745 Good example
2022-09-14 14:07:45,745 Training model without error in pos_label
2022-09-14 14:07:45,745 ...unique values in y_train ['Negative_label' 'Positive_label']
2022/09/14 14:07:47 INFO mlflow.projects: === Run (ID 'ec5922588def43dea40f502f5482245a') succeeded ===
```

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by harupy** [comment]
@nelsoncardenas Thanks for reporting the issue. It looks like `pos_label` is not used at all when creating precision-recall curve:

https://github.com/mlflow/mlflow/blob/f7b19b8dcacff24ddfa535e7a9808cb51bdf96ea/mlflow/sklearn/utils.py#L373-L385

**Comment by harupy** [comment]
I think we can just make the following change:

```diff
diff --git a/mlflow/sklearn/utils.py b/mlflow/sklearn/utils.py
index 604407112..b94c1acfd 100644
--- a/mlflow/sklearn/utils.py
+++ b/mlflow/sklearn/utils.py
@@ -280,7 +280,7 @@ def _get_class_labels_from_estimator(estimator):
     return estimator.classes_ if hasattr(estimator, "classes_") else None
 
 
-def _get_classifier_artifacts(fitted_estimator, prefix, X, y_true, sample_weight):
+def _get_classifier_artifacts(fitted_estimator, prefix, X, y_true, sample_weight, pos_label):
     """
     Draw and record various common artifacts for classifier
 
@@ -380,6 +380,7 @@ def _get_classifier_artifacts(fitted_estimator, prefix, X, y_true, sample_weight
                         X=X,
                         y=y_true,
                         sample_weight=sample_weight,
+                        pos_label=pos_label,
                     ),
                     title="Precision recall curve",
                 ),
@@ -513,7 +514,7 @@ def _log_specialized_estimator_content(
     if sklearn.base.is_classifier(fitted_estimator):
         try:
             artifacts = _get_classifier_artifacts(
-                fitted_estimator, prefix, X, y_true, sample_weight
+                fitted_estimator, prefix, X, y_true, sample_weight, pos_label
             )
         except Exception as e:
             msg = (
```

**Comment by harupy** [comment]
We might need to pass `pos_label` to other plotting functions.

**Comment by harupy** [comment]
@nelsoncardenas Would you be interested in working on fixing this issue?

**Comment by nelsoncardenas** [comment]
@harupy I'd like to help, I just want to make sure I'll follow the right guidelines from you, this is the first time trying to contribute to MLflow

**Comment by nelsoncardenas** [comment]
Actually, I'm reading the guideline and I found another error, do you think I should create a new issue? 

https://github.com/mlflow/mlflow/blob/0a7760107ef208efcfbba35f9b74659683b239d2/CONTRIBUTING.md?plain=1#L510

**Comment by harupy** [comment]
@nelsoncardenas It looks like https://github.com/mlflow/mlflow/pull/5717/ renamed the `Common prerequisites and dependencies` section. Feel free to file a new issue or a PR to fix it :)

**Comment by nelsoncardenas** [comment]
@harupy sorry, I think I'll not be able to work on this issue, right now I don't have so much time available.

**Comment by immu4989** [comment]
I'd like to pick this up — it's still reproducible on master.

The metrics path was since fixed: _get_classifier_metrics now threads pos_label (mlflow/sklearn/utils.py, used at the _log_specialized_estimator_content call site). However, the artifact path was not — _get_classifier_artifacts has no pos_label parameter, and the RocCurveDisplay.from_estimator / PrecisionRecallDisplay.from_estimator argument dicts don't pass one. So for a binary classifier whose positive class isn't classes_[1], the autologged ROC and precision-recall curves are still computed for the wrong class even though the scalar metrics are now correct.

**Proposed fix:** add pos_label to _get_classifier_artifacts(...), pass it from _log_specialized_estimator_content (the variable is already in scope there), and include "pos_label": pos_label in the ROC/PR curve argument dicts. I'll add a regression test asserting the curve is built for the configured positive label. 

Will open a PR if that direction looks good.

**Comment by vedjaw** [comment]
Opened PR # with the full MLflow PR template. Fix threads `pos_label` through `_get_classifier_artifacts()` into ROC/PR plot calls, with regression tests.

**Comment by immu4989** [comment]
@vedjaw , I'll be honest that it's a bit disappointing to see this implemented without any acknowledgment. I posted the full root cause analysis and the exact fix direction — threading pos_label through _get_classifier_artifacts() into the RocCurveDisplay / PrecisionRecallDisplay calls — in [my comment above](https://github.com/mlflow/mlflow/issues/6797#issuecomment-4584886582) two weeks ago, along with a prepared fix and regression tests, and was simply waiting on the ready label before opening a PR.

#23977 mirrors that analysis closely. Building on someone's public diagnosis is completely fine and welcome , that's how open source works , but a quick acknowledgment when you do is the usual courtesy, and it would have been appreciated here.
